lib/train.rb

```diff
@@ -61,53 +61,52 @@ def self.load_transport(transport_name)
     raise ex
   end
 
-  # Resolve target configuration in URI-scheme into
-  # all respective fields and merge with existing configuration.
-  # e.g. ssh://bob@remote  =>  backend: ssh, user: bob, host: remote
-  def self.target_config(config = nil) # rubocop:disable Metrics/AbcSize
-    conf = config.nil? ? {} : config.dup
-    conf = symbolize_keys(conf)
-
-    group_keys_and_keyfiles(conf)
+  # Legacy code to unpack a series of items from an incoming Hash
+  # Inspec::Config.unpack_train_credentials now handles this in most cases that InSpec needs
+  # If you need to unpack a URI, use unpack_target_from_uri
+  def self.target_config(config = nil)
+    conf = config.dup
+    # Symbolize keys
+    conf.keys.each do |key|
+      unless key.is_a? Symbol
+        conf[key.to_sym] = conf.delete(key)
+      end
+    end
 
+    group_keys_and_keyfiles(conf) # TODO: move logic into SSH plugin
     return conf if conf[:target].to_s.empty?
+    unpack_target_from_uri(conf[:target], conf).merge(conf)
+  end
+
+  def self.unpack_target_from_uri(uri_string, opts = {}) # rubocop: disable Metrics/AbcSize
+    creds = {}
+    return creds if uri_string.empty?
 
     # split up the target's host/scheme configuration
-    uri = parse_uri(conf[:target].to_s)
+    uri = parse_uri(uri_string)
     unless uri.host.nil? and uri.scheme.nil?
-      conf[:backend]  ||= uri.scheme
-      conf[:host]     ||= uri.hostname
-      conf[:port]     ||= uri.port
-      conf[:user]     ||= uri.user
-      conf[:path]     ||= uri.path
-      conf[:password] ||=
-        if conf[:www_form_encoded_password] && !uri.password.nil?
+      creds[:backend]  ||= uri.scheme
+      creds[:host]     ||= uri.hostname
+      creds[:port]     ||= uri.port
+      creds[:user]     ||= uri.user
+      creds[:path]     ||= uri.path
+      creds[:password] ||=
+        if opts[:www_form_encoded_password] && !uri.password.nil?
           URI.decode_www_form_component(uri.password)
         else
           uri.password
         end
     end
 
-    # ensure path is nil, if its empty; e.g. required to reset defaults for winrm
-    conf[:path] = nil if !conf[:path].nil? && conf[:path].to_s.empty?
+    # ensure path is nil, if its empty; e.g. required to reset defaults for winrm # TODO: move logic into winrm plugin
+    creds[:path] = nil if !creds[:path].nil? && creds[:path].to_s.empty?
 
-    # return the updated config
-    conf
-  end
+    # compat! is available in ruby 2.4+
```

Comment:
Could/Should we add a TODO line to change this? If not, any reason for this comment?